### PR TITLE
Gc threading

### PR DESCRIPTION
**Pull request by dvirsky**
This PR makes the garbage collector work using a side thread for each spec, changing the collection frequency dynamically. For the design see docs/design/gc.md below.

### CODE REVIEW

File: docs/design/gc.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# Garbage Collection in RediSearch
```
Comment (dvirsky) [inline docs/design/gc.md]: TODO: Link to this from the website or something

File: src/buffer.c [hunk 2]
```diff
@@ -31,7 +31,12 @@ size_t Buffer_Truncate(Buffer *b, size_t newlen) {
     newlen = Buffer_Offset(b);
   }
 
-  b->data = rm_realloc(b->data, newlen);
+  if (newlen == 0) {
```
Comment (dvirsky) [inline src/buffer.c]: This is because with the GC we can now have empty blocks.

File: src/config.c [hunk 3]
```diff
@@ -18,5 +18,9 @@ int ReadConfig(RedisModuleString **argv, int argc, const char **err) {
     RSGlobalConfig.concurrentMode = 0;
   }
 
+  if (RMUtil_ArgIndex("NOGC", argv, argc) >= 0) {
```
Comment (dvirsky) [inline src/config.c]: needs documentation

File: src/gc.c [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+GarbageCollectorCtx *NewGarbageCollector(const RedisModuleString *k, float initialHZ) {
+  GarbageCollectorCtx *gc = malloc(sizeof(*gc));
+
+  *gc = (GarbageCollectorCtx){
+      .timer = NULL, .hz = initialHZ, .keyName = k, .stats = {},
+  };
+  return gc;
+}
+
+int isRdbLoading(RedisModuleCtx *ctx) {
```
Comment (dvirsky) [inline src/gc.c]: worth moving to RMUtil

File: src/gc.c [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+  if (!RMUtilInfo_GetInt(info, "loading", &isLoading)) {
+    isLoading = 0;
+  }
+
+  RMUtilRedisInfo_Free(info);
+  return isLoading == 1;
+}
+
+void GC_PeriodicCallback(RedisModuleCtx *ctx, void *privdata) {
```
Comment (dvirsky) [inline src/gc.c]: needs more inline comments to explain what's going on

File: src/gc.c [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+end:
+  if (sctx) {
+    RedisModule_CloseKey(sctx->key);
+    SearchCtx_Free(sctx);
+  }
+  if (idxKey) RedisModule_CloseKey(idxKey);
+  RedisModule_ThreadSafeContextUnlock(ctx);
+}
+
+static void gc_onTerm(void *privdata) {
```
Comment (dvirsky) [inline src/gc.c]: needs documentation, why is this here?

File: src/gc.c [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+// Start the collector thread
+int GC_Start(GarbageCollectorCtx *ctx) {
+  assert(ctx->timer == NULL);
+  ctx->timer = RMUtil_NewPeriodicTimer(GC_PeriodicCallback, gc_onTerm, ctx, hzToTimeSpec(ctx->hz));
+  return REDISMODULE_OK;
+}
+
+int GC_Stop(GarbageCollectorCtx *ctx) {
```
Comment (dvirsky) [inline src/gc.c]: needs docs

File: src/gc.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+typedef struct GarbageCollectorCtx GarbageCollectorCtx;
+
+GarbageCollectorCtx *NewGarbageCollector(const RedisModuleString *k, float initial_hz);
+
+// Start the collector thread
+int GC_Start(GarbageCollectorCtx *ctx);
+
+int GC_Stop(GarbageCollectorCtx *ctx);
+
+void GC_Free(GarbageCollectorCtx *ctx);
```
Comment (dvirsky) [inline src/gc.h]: the implementation was removed, need to remove the declaration

File: src/inverted_index.h [hunk 9]
```diff
@@ -26,6 +25,7 @@ typedef struct {
   IndexFlags flags;
   t_docId lastId;
   uint32_t numDocs;
+  uint32_t gcMarker;
```
Comment (dvirsky) [inline src/inverted_index.h]: needs an explanation here

File: src/module.c [hunk 10]
```diff
@@ -473,7 +424,9 @@ int DeleteCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
   int rc = DocTable_Delete(&sp->docs, RedisModule_StringPtrLen(argv[2], NULL));
   if (rc == 1) {
     sp->stats.numDocuments--;
+    GC_OnDelete(sp->gc);
```
Comment (dvirsky) [inline src/module.c]: explain this...

File: src/module.c [hunk 11]
```diff
@@ -558,7 +511,8 @@ int AddHashCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
   doc.payloadSize = 0;
   Document_Detach(&doc, ctx);
 
-  RedisModuleBlockedClient *client = RedisModule_BlockClient(ctx, NULL, NULL, NULL, 0);
+  RedisModuleBlockedClient *client =
```
Comment (dvirsky) [inline src/module.c]: fixed unrelated memory leak I found while testing this branch with valgrind

File: src/redis_index.c [hunk 12]
```diff
@@ -34,8 +34,10 @@ void *InvertedIndex_RdbLoad(RedisModuleIO *rdb, int encver) {
 
     size_t cap;
     char *data = RedisModule_LoadStringBuffer(rdb, &cap);
-    blk->data = Buffer_Wrap(data, cap);
+
+    blk->data = Buffer_Wrap(cap > 0 ? data : NULL, cap);
     blk->data->offset = cap;
+    if (!cap && data) RedisModule_Free(data);
```
Comment (dvirsky) [inline src/redis_index.c]: explain...

File: src/redis_index.c [hunk 13]
```diff
@@ -148,7 +150,9 @@ RedisSearchCtx *NewSearchCtx(RedisModuleCtx *ctx, RedisModuleString *indexName)
 }
 
 void SearchCtx_Free(RedisSearchCtx *sctx) {
-
+  // if (sctx->keyName && sctx->redisCtx) {
```
Comment (dvirsky) [inline src/redis_index.c]: remove

File: src/redis_index.c [hunk 14]
```diff
@@ -224,7 +228,8 @@ const char *Redis_SelectRandomTerm(RedisSearchCtx *ctx, size_t *tlen) {
   return NULL;
 }
 // ScoreIndex *LoadRedisScoreIndex(RedisSearchCtx *ctx, const char *term, size_t len) {
-//   Buffer *b = NewRedisBuffer(ctx->redisCtx, fmtRedisScoreIndexKey(ctx, term, len), BUFFER_READ);
+//   Buffer *b = NewRedisBuffer(ctx->redisCtx, fmtRedisScoreIndexKey(ctx, term, len),
```
Comment (dvirsky) [inline src/redis_index.c]: remove all this crap

File: src/rmutil/periodic.c [hunk 15]
```diff
@@ -5,6 +5,7 @@
 
 typedef struct RMUtilTimer {
   RMutilTimerFunc cb;
+  RMUtilTimerTerminationFunc onTerm;
```
Comment (dvirsky) [inline src/rmutil/periodic.c]: a lot of changes to periodic timer, we need to port them back to rmutil

File: src/spec.c [hunk 16]
```diff
@@ -256,9 +262,63 @@ void IndexSpec_RestoreTerm(IndexSpec *sp, const char *term, size_t len, double s
...
+  for (size_t i = 0; i < len; i++) {
+    if (selection >= totalWeight && selection <= (totalWeight + weights[i])) {
+      return i;
+    }
+    totalWeight += weights[i];
+  }
+  // fallback
+  return 0;
+}
+char *IndexSpec_GetRandomTerm(IndexSpec *sp, size_t sampleSize) {
```
Comment (dvirsky) [inline src/spec.c]: Document this

File: src/spec.c [hunk 17]
```diff
@@ -356,10 +416,22 @@ IndexSpec *NewIndexSpec(const char *name, size_t numFields) {
   sp->stopwords = DefaultStopWordList();
   sp->terms = NewTrie();
   sp->sortables = NULL;
+  sp->gc = NULL;
   memset(&sp->stats, 0, sizeof(sp->stats));
   return sp;
 }
 
+void IndexSpec_StartGC(RedisModuleCtx *ctx, IndexSpec *sp, float initialHZ) {
```
Comment (dvirsky) [inline src/spec.c]: Document this

File: src/tests/test_trie.c [hunk 18]
```diff
@@ -178,6 +178,16 @@ int testTrie() {
   sc = TrieNode_Find(root, runes, rlen);
   ASSERT(sc == 12);
 
+  double score;
```
Comment (dvirsky) [inline src/tests/test_trie.c]: needs a better test

File: src/trie/trie_type.c [hunk 19]
```diff
@@ -238,6 +238,29 @@ Vector *Trie_Search(Trie *tree, char *s, size_t len, size_t num, int maxDist, in
   return ret;
 }
 
+int Trie_RandomKey(Trie *t, char **str, t_len *len, double *score) {
```
Comment (dvirsky) [inline src/trie/trie_type.c]: needs documentation an inline comments

File: src/spec.c [hunk 20]
```diff
@@ -62,7 +64,11 @@ IndexSpec *IndexSpec_ParseRedisArgs(RedisModuleCtx *ctx, RedisModuleString *name
     args[i] = RedisModule_StringPtrLen(argv[i], NULL);
   }
 
-  return IndexSpec_Parse(RedisModule_StringPtrLen(name, NULL), args, argc, err);
+  IndexSpec *ret = IndexSpec_Parse(RedisModule_StringPtrLen(name, NULL), args, argc, err);
+  if (ret) {
+    IndexSpec_StartGC(ctx, ret, GC_DEFAULT_HZ);
```
Comment (mnunberg) [inline src/spec.c]: This looks really dirty. Maybe you should abstract this when a new IndexSpec is created?

Comment (dvirsky) [inline src/spec.c]: This is where it gets created, The name of the function might be misleading though.

File: src/gc.c (hunk 5 above)
Comment (mnunberg) [inline src/gc.c]: `static`

File: src/gc.c [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+      // blockNum 0 means error or we've finished
+      if (!blockNum) break;
+
+      // After each iteration we yield execution
+      // First we close the relevant keys we're touching
+      RedisModule_CloseKey(sctx->key);
+      RedisModule_CloseKey(idxKey);
+      SearchCtx_Free(sctx);
+
+      // now release the global lock
```
Comment (mnunberg) [inline src/gc.c]: Don't we have a YIELD macro for this?
On second thought, the whole reopening logic might be better done in a function.

File: src/gc.c (hunk 6 above)
Comment (mnunberg) [inline src/gc.c]: Might help to have these 'contextless' functions in RMUtil... (i.e. for freeing memory - this way it might be able to have a static context and simply free it using that).

File: src/gc.h [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+
+// get the current stats from the collector
+struct GCStats *GC_GetStats(GarbageCollectorCtx *ctx);
```
Comment (mnunberg) [inline src/gc.h]: `const`?

File: src/inverted_index.c [hunk 23]
```diff
@@ -76,9 +76,26 @@ void IndexReader_OnReopen(RedisModuleKey *k, void *privdata) {
...
   ir->idx = RedisModule_ModuleTypeGetValue(k);
-  size_t offset = ir->br.pos;
-  ir->br = NewBufferReader(IR_CURRENT_BLOCK(ir).data);
-  ir->br.pos = offset;
+
+  // the gc marker tells us if there is a chance the keys has undergone GC while we were asleep
+  if (ir->gcMarker == ir->idx->gcMarker) {
+    // no GC - we just go to the same offset we were at
+    size_t offset = ir->br.pos;
+    ir->br = NewBufferReader(IR_CURRENT_BLOCK(ir).data);
```
Comment (mnunberg) [inline src/inverted_index.c]: This line and line 92 (`NewBufferReader(...)`) have the same implication. They should be outside?

File: src/inverted_index.c [hunk 24]
```diff
@@ -404,7 +420,7 @@ int IR_Read(void *ctx, RSIndexResult **e) {
     goto eof;
   }
   do {
-    if (BufferReader_AtEnd(&ir->br)) {
+    while (BufferReader_AtEnd(&ir->br)) {
```
Comment (mnunberg) [inline src/inverted_index.c]: nested `while` inside `do` looks weird. Can it be moved to a different function or macro?

File: src/inverted_index.c [hunk 25]
```diff
@@ -646,42 +667,58 @@ int IndexBlock_Repair(IndexBlock *blk, DocTable *dt, IndexFlags flags) {
...
 
-    if (md->flags & Document_Deleted) {
-      frags += 1;
-      // printf("ignoring hole in doc %d, frags now %d\n", docId, frags);
-    } else {
+    // If we found a deleted document, we increment the number of found "frags",
+    // and not write anything, so the reader will advance byt the writer won't.
+    // this will close the "hole" in the index
+    if (!md || md->flags & Document_Deleted) {
+      ++frags;
```
Comment (mnunberg) [inline src/inverted_index.c]: Nice to know it's not Python ;)

File: src/inverted_index.c [hunk 26]
```diff
@@ -646,42 +667,58 @@ int IndexBlock_Repair(IndexBlock *blk, DocTable *dt, IndexFlags flags) {
     lastReadId = res->docId += lastReadId;
     RSDocumentMetadata *md = DocTable_Get(dt, res->docId);
 
-    if (md->flags & Document_Deleted) {
-      frags += 1;
-      // printf("ignoring hole in doc %d, frags now %d\n", docId, frags);
-    } else {
+    // If we found a deleted document, we increment the number of found "frags",
+    // and not write anything, so the reader will advance byt the writer won't.
```
Comment (mnunberg) [inline src/inverted_index.c]: byt -> but?

File: src/gc.c (hunk 6 above)
Comment (dvirsky) [inline src/gc.c]: well, you can always pass `free()` to to the timer ;)

File: src/gc.c (hunk 21 above)
Comment (dvirsky) [inline src/gc.c]: I wanted this disconnected from the concurrent context

File: src/inverted_index.c (hunk 24 above)
Comment (dvirsky) [inline src/inverted_index.c]: i don't want a function as this is the only place it happens. The reason this shifted from an if to while, is that we might have empty index blocks, so assuming we're at the end and the next block is valid is not true anymore.

File: src/inverted_index.c (hunk 23 above)
Comment (dvirsky) [inline src/inverted_index.c]: I moved it, not sure if it doesn't just make the code less readable.

File: src/spec.c (hunk 20 above)
Comment (dvirsky) [inline src/spec.c]: I've created a new function for creation that makes things clearer

File: src/inverted_index.c (hunk 23 above)
Comment (dvirsky) [inline src/inverted_index.c]: Oh, it actually creates a bug, there was a reason for it! :) we have to do the `size_t offset = ir->br.pos;` thing before that.
